File: src/retrieval/conflict.ts
import type {
  MemoryEntry,
  ConflictResult,
  Embedder,
  StorageBackend,
} from '../core/types.js';
import { cosineSimilarity } from '../core/utils.js';

/** Words/phrases that signal negation or opposition */
const NEGATION_PATTERNS = [
  /\bnot\b/i, /\bnever\b/i, /\bno\b/i, /\bdon't\b/i, /\bdoesn't\b/i,
  /\bwon't\b/i, /\bcan't\b/i, /\bhate\b/i, /\bdislike\b/i, /\bavoid\b/i,
  /\bstop\b/i, /\bquit\b/i, /\bremove\b/i, /\bdelete\b/i,
  /\binstead of\b/i, /\brather than\b/i, /\bno longer\b/i,
];

/** Words that signal preference changes */
const CHANGE_PATTERNS = [
  /\bnow\b/i, /\bactually\b/i, /\bchanged?\b/i, /\bswitch/i,
  /\bprefer\b/i, /\bwant\b/i, /\bused to\b/i, /\banymore\b/i,
];

/**
 * Conflict detector.
 * Compares new content against stored memories to find contradictions.
 * Uses semantic similarity + negation detection, not just keyword matching.
 */
export class ConflictDetector {
  private store: StorageBackend;
  private embedder: Embedder;
  /** Similarity threshold to consider two entries as "same topic" */
  private topicThreshold: number;

  constructor(
    store: StorageBackend,
    embedder: Embedder,
    topicThreshold = 0.6
  ) {
    this.store = store;
    this.embedder = embedder;
    this.topicThreshold = topicThreshold;
  }

  /**
   * Check if new content conflicts with any stored memories.
   * Returns conflicts sorted by confidence (highest first).
   */
  async check(content: string, scope?: string): Promise<ConflictResult[]> {
    const entries = await this.store.getAll(scope);
    if (entries.length === 0) return [];

    const contentEmbedding = await this.embedder.embed(content);
    const conflicts: ConflictResult[] = [];

    for (const entry of entries) {
      const entryEmbedding = entry.embedding ?? await this.embedder.embed(entry.content);
      const similarity = cosineSimilarity(contentEmbedding, entryEmbedding);

      // Only check entries on the same topic
      if (similarity < this.topicThreshold) continue;

      // Check for contradiction signals
      const contradictionScore = this.scoreContradiction(content, entry.content);
      if (contradictionScore <= 0) continue;

      // Combined confidence: high topic similarity + contradiction signals
      const confidence = Math.min(1, similarity * 0.5 + contradictionScore * 0.5);

      conflicts.push({
        incoming: content,
        stored: entry,
        confidence,
        action: this.suggestAction(entry, confidence),
        reason: this.generateReason(content, entry, contradictionScore),
      });
    }

    conflicts.sort((a, b) => b.confidence - a.confidence);
    return conflicts;
  }

  /**
   * Score how contradictory two pieces of text are.
   * Returns 0 (no contradiction) to 1 (strong contradiction).
   */
  private scoreContradiction(incoming: string, stored: string): number {
    let score = 0;

    // Check for negation in incoming that opposes stored
    const incomingHasNegation = NEGATION_PATTERNS.some(p => p.test(incoming));
    const storedHasNegation = NEGATION_PATTERNS.some(p => p.test(stored));

    // One is negated, the other isn't - likely contradiction
    if (incomingHasNegation !== storedHasNegation) {
      score += 0.5;
    }

    // Check for explicit change language
    const hasChangeLanguage = CHANGE_PATTERNS.some(p => p.test(incoming));
    if (hasChangeLanguage) {
      score += 0.3;
    }

    // Check for antonym pairs in the texts
    const antonymScore = this.checkAntonyms(incoming, stored);
    score += antonymScore * 0.4;

    return Math.min(1, score);
  }

  /** Check for common antonym pairs */
  private checkAntonyms(a: string, b: string): number {
    const antonyms: [RegExp, RegExp][] = [
      [/\blike\b/i, /\bdislike\b/i],
      [/\blove\b/i, /\bhate\b/i],
      [/\byes\b/i, /\bno\b/i],
      [/\btrue\b/i, /\bfalse\b/i],
      [/\bvegetarian\b/i, /\bmeat\b/i],
      [/\bvegan\b/i, /\bmeat\b|\bdairy\b/i],
      [/\bmorning\b/i, /\bevening\b|\bnight\b/i],
      [/\blight\b/i, /\bdark\b/i],
      [/\bhot\b/i, /\bcold\b/i],
      [/\bfast\b/i, /\bslow\b/i],
      [/\benable\b/i, /\bdisable\b/i],
      [/\ballow\b/i, /\bblock\b|\bdeny\b/i],
      [/\baccept\b/i, /\breject\b/i],
    ];

    let matches = 0;
    for (const [word1, word2] of antonyms) {
      if ((word1.test(a) && word2.test(b)) || (word2.test(a) && word1.test(b))) {
        matches++;
      }
    }
    return Math.min(1, matches * 0.5);
  }

  private suggestAction(stored: MemoryEntry, confidence: number): ConflictResult['action'] {
    // Hard constraints should always trigger clarification
    if (stored.importance === 'hard') return 'clarify';
    // High confidence conflicts with soft importance - suggest override
    if (confidence > 0.7 && stored.importance === 'soft') return 'override';
    // Ephemeral memories can be overridden easily
    if (stored.importance === 'ephemeral') return 'override';
    // Default: ask for clarification
    return 'clarify';
  }

  private generateReason(
    incoming: string,
    stored: MemoryEntry,
    contradictionScore: number
  ): string {
    if (stored.importance === 'hard') {
      return `Conflicts with a hard constraint: "${stored.content}". This memory was marked as non-negotiable - please confirm the change.`;
    }
    if (contradictionScore > 0.7) {
      return `Directly contradicts stored memory: "${stored.content}". The new statement appears to reverse a previous preference.`;
    }
    return `Potentially conflicts with: "${stored.content}". The statements may be inconsistent.`;
  }
}

File: python/agentic_memory/retrieval/conflict.py
"""Conflict detector for finding contradictions in stored memories."""

from __future__ import annotations

import re
from typing import Any, Optional

from ..types import MemoryEntry, ConflictResult, ConflictAction, ImportanceLevel
from ..utils import cosine_similarity

NEGATION_PATTERNS = [
    re.compile(p, re.IGNORECASE) for p in [
        r"\bnot\b", r"\bnever\b", r"\bno\b", r"\bdon't\b", r"\bdoesn't\b",
        r"\bwon't\b", r"\bcan't\b", r"\bhate\b", r"\bdislike\b", r"\bavoid\b",
        r"\bstop\b", r"\bquit\b", r"\bremove\b", r"\bdelete\b",
        r"\binstead of\b", r"\brather than\b", r"\bno longer\b",
    ]
]

CHANGE_PATTERNS = [
    re.compile(p, re.IGNORECASE) for p in [
        r"\bnow\b", r"\bactually\b", r"\bchanged?\b", r"\bswitch",
        r"\bprefer\b", r"\bwant\b", r"\bused to\b", r"\banymore\b",
    ]
]

ANTONYM_PAIRS: list[tuple[re.Pattern, re.Pattern]] = [
    (re.compile(a, re.IGNORECASE), re.compile(b, re.IGNORECASE))
    for a, b in [
        (r"\blike\b", r"\bdislike\b"),
        (r"\blove\b", r"\bhate\b"),
        (r"\byes\b", r"\bno\b"),
        (r"\btrue\b", r"\bfalse\b"),
        (r"\bvegetarian\b", r"\bmeat\b"),
        (r"\bvegan\b", r"\bmeat\b|\bdairy\b"),
        (r"\bmorning\b", r"\bevening\b|\bnight\b"),
        (r"\blight\b", r"\bdark\b"),
        (r"\bhot\b", r"\bcold\b"),
        (r"\bfast\b", r"\bslow\b"),
        (r"\benable\b", r"\bdisable\b"),
        (r"\ballow\b", r"\bblock\b|\bdeny\b"),
        (r"\baccept\b", r"\breject\b"),
    ]
]


class ConflictDetector:
    """
    Conflict detector.
    Compares new content against stored memories to find contradictions.
    Uses semantic similarity + negation detection.
    """

    def __init__(
        self,
        store: Any,
        embedder: Any,
        topic_threshold: float = 0.6,
    ) -> None:
        self._store = store
        self._embedder = embedder
        self._topic_threshold = topic_threshold

    async def check(self, content: str, scope: Optional[str] = None) -> list[ConflictResult]:
        entries = await self._store.get_all(scope)
        if not entries:
            return []

        content_embedding = await self._embedder.embed(content)
        conflicts: list[ConflictResult] = []

        for entry in entries:
            entry_embedding = entry.embedding or await self._embedder.embed(entry.content)
            similarity = cosine_similarity(content_embedding, entry_embedding)

            if similarity < self._topic_threshold:
                continue

            contradiction_score = self._score_contradiction(content, entry.content)
            if contradiction_score <= 0:
                continue

            confidence = min(1.0, similarity * 0.5 + contradiction_score * 0.5)

            conflicts.append(ConflictResult(
                incoming=content,
                stored=entry,
                confidence=confidence,
                action=self._suggest_action(entry, confidence),
                reason=self._generate_reason(content, entry, contradiction_score),
            ))

        conflicts.sort(key=lambda c: c.confidence, reverse=True)
        return conflicts

    def _score_contradiction(self, incoming: str, stored: str) -> float:
        score = 0.0

        incoming_negated = any(p.search(incoming) for p in NEGATION_PATTERNS)
        stored_negated = any(p.search(stored) for p in NEGATION_PATTERNS)

        if incoming_negated != stored_negated:
            score += 0.5

        if any(p.search(incoming) for p in CHANGE_PATTERNS):
            score += 0.3

        score += self._check_antonyms(incoming, stored) * 0.4

        return min(1.0, score)

    @staticmethod
    def _check_antonyms(a: str, b: str) -> float:
        matches = 0
        for word1, word2 in ANTONYM_PAIRS:
            if (word1.search(a) and word2.search(b)) or (word2.search(a) and word1.search(b)):
                matches += 1
        return min(1.0, matches * 0.5)

    @staticmethod
    def _suggest_action(stored: MemoryEntry, confidence: float) -> ConflictAction:
        if stored.importance == ImportanceLevel.HARD:
            return ConflictAction.CLARIFY
        if confidence > 0.7 and stored.importance == ImportanceLevel.SOFT:
            return ConflictAction.OVERRIDE
        if stored.importance == ImportanceLevel.EPHEMERAL:
            return ConflictAction.OVERRIDE
        return ConflictAction.CLARIFY

    @staticmethod
    def _generate_reason(incoming: str, stored: MemoryEntry, contradiction_score: float) -> str:
        if stored.importance == ImportanceLevel.HARD:
            return (
                f'Conflicts with a hard constraint: "{stored.content}". '
                "This memory was marked as non-negotiable - please confirm the change."
            )
        if contradiction_score > 0.7:
            return (
                f'Directly contradicts stored memory: "{stored.content}". '
                "The new statement appears to reverse a previous preference."
            )
        return (
            f'Potentially conflicts with: "{stored.content}". '
            "The statements may be inconsistent."
        )
